From f9da0329b6346178355e6337215882e1a0cc0369 Mon Sep 17 00:00:00 2001
From: Brecht Van Lommel <brecht@blender.org>
Date: Fri, 10 May 2024 13:32:53 +0200
Subject: [PATCH] BLENDER: Workaround LFS files not being available in pull
 requests

Patch taken from issue number 17715, associating the LFS pointer when the
file is downloaded.
---
 models/git/lfs.go      |  2 +-
 services/lfs/server.go | 20 ++++++++++++++++++++
 2 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/models/git/lfs.go b/models/git/lfs.go
index 837dc9fd312f5..b52208a77f5d8 100644
--- a/models/git/lfs.go
+++ b/models/git/lfs.go
@@ -238,7 +238,7 @@ func CountLFSMetaObjects(ctx context.Context, repoID int64) (int64, error) {

 // LFSObjectAccessible checks if a provided Oid is accessible to the user
 func LFSObjectAccessible(ctx context.Context, user *user_model.User, oid string) (bool, error) {
-	if user.IsAdmin {
+	if user != nil && user.IsAdmin {
 		count, err := db.GetEngine(ctx).Count(&LFSMetaObject{Pointer: lfs.Pointer{Oid: oid}})
 		return count > 0, err
 	}
diff --git a/services/lfs/server.go b/services/lfs/server.go
index ace501e15f0df..e5873e8c38a86 100644
--- a/services/lfs/server.go
+++ b/services/lfs/server.go
@@ -246,6 +246,26 @@ func BatchHandler(ctx *context.Context) {
 			responseObject = buildObjectResponse(rc, p, false, !exists, err)
 		} else {
 			var err *lfs_module.ObjectError
+
+			if exists && meta == nil {
+				accessible, accessibleErr := git_model.LFSObjectAccessible(ctx, ctx.Doer, p.Oid)
+				if accessibleErr != nil {
+					log.Error("Unable to check if LFS MetaObject [%s] is accessible. Error: %v", p.Oid, err)
+					writeStatus(ctx, http.StatusInternalServerError)
+					return
+				}
+				if accessible {
+					_, newMetaObjErr := git_model.NewLFSMetaObject(ctx, repository.ID, p)
+					if newMetaObjErr != nil {
+						log.Error("Unable to create LFS MetaObject [%s] for %s/%s. Error: %v", p.Oid, rc.User, rc.Repo, err)
+						writeStatus(ctx, http.StatusInternalServerError)
+						return
+					}
+				} else {
+					exists = false
+				}
+			}
+
 			if !exists || meta == nil {
 				err = &lfs_module.ObjectError{
 					Code:    http.StatusNotFound,
